common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java

```diff
@@ -0,0 +1,814 @@
+            }
+
+            if (lootrun.notes().containsKey(chunkLong)) {
+                renderNotes(event.getPoseStack(), lootrun, color, source, chunkLong);
+            }
+        }
+
+        event.getPoseStack().popPose();
+    }
+
+    private static void renderNotes(
+            PoseStack poseStack,
+            LootrunInstance lootrun,
+            int color,
+            MultiBufferSource.BufferSource source,
+            long chunkLong) {
+        List<Pair<Vec3, Component>> notes = lootrun.notes().get(chunkLong);
+
+        Font font = McUtils.mc().font;
+
+        for (Pair<Vec3, Component> note : notes) {
+            Vec3 location = note.first();
+            poseStack.pushPose();
+            poseStack.translate(location.x, location.y + 2, location.z);
+            poseStack.mulPose(McUtils.mc().gameRenderer.getMainCamera().rotation());
+            poseStack.scale(-0.025f, -0.025f, 0.025f);
+            Matrix4f pose = poseStack.last().pose();
+            List<FormattedCharSequence> lines = font.split(note.right(), 200);
+            int offsetY = -(font.lineHeight * lines.size()) / 2;
+            for (FormattedCharSequence line : lines) {
+                int offsetX = -font.width(line) / 2;
+                font.drawInBatch(line, offsetX, offsetY, color, false, pose, source, false, 0x80000000, 0xf000f0);
+                offsetY += font.lineHeight + 2;
+            }
+            poseStack.popPose();
+        }
+    }
+
+    private static void renderChests(
+            PoseStack poseStack,
+            LootrunInstance lootrun,
+            int color,
+            MultiBufferSource.BufferSource source,
+            long chunkLong) {
+        VertexConsumer consumer = source.getBuffer(RenderType.lines());
+        Set<BlockPos> chests = lootrun.chests().get(chunkLong);
+
+        float red = ((float) FastColor.ARGB32.red(color)) / 255;
+        float green = ((float) FastColor.ARGB32.green(color)) / 255;
+        float blue = ((float) FastColor.ARGB32.blue(color)) / 255;
+
+        for (BlockPos chest : chests) {
+            LevelRenderer.renderLineBox(poseStack, consumer, new AABB(chest), red, green, blue, 1f);
+        }
+
+        source.endBatch();
+    }
+
+    private static void renderPoints(
+            PoseStack poseStack,
+            MultiBufferSource.BufferSource source,
+            Long2ObjectMap<List<List<ColoredPoint>>> points,
+            long chunkLong) {
+        List<List<ColoredPoint>> locations = points.get(chunkLong);
+
+        Level level = McUtils.mc().level;
+        if (level == null) return;
+
+        for (List<ColoredPoint> locationsInRoute : locations) {
+            VertexConsumer consumer = source.getBuffer(LootrunRenderType.LOOTRUN_LINE);
+            Matrix4f lastMatrix = poseStack.last().pose();
+            boolean sourceBatchEnded = false;
+
+            List<ColoredPoint> toRender = new ArrayList<>();
```

Comment:
`toRender` seems to be slightly convoluted in its implementation, and seems to be a way of storing positions that are in the air or equal to last block position?? for later rendering.

Reply:
Points are collected every tick, so they happen to be the same block a lot of the times. That's why the caching exists, so it does not recalculate block validness.